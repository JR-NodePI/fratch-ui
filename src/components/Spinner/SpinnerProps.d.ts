import { SpinnerType } from './SpinnerConstants';

export type SpinnerProps = {
  cover?: boolean;
  inverted?: boolean;
  label?: string;
  type?: (typeof SpinnerType)[keyof typeof SpinnerType];
};
