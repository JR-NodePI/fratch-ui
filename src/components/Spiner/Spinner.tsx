import { c } from "../../helpers/classNameHelpers";

import styles from "./Spinner.module.css";

const Spinner = (): JSX.Element => {
  return (
    <svg className={c(styles.spinner)} viewBox="0 0 50 50">
      <circle
        className={c(styles.spinner_path)}
        cx="25"
        cy="25"
        r="20"
        fill="none"
        stroke-width="5"
      ></circle>
    </svg>
  );
};

export default Spinner;
