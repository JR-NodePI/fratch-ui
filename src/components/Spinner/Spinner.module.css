.spinner_container {
  position: relative;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}
.spinner_container.cover {
  position: fixed;
  z-index: 999;
  top: 0;
  left: 0;
  background-color: rgba(255, 255, 255, 0.8);
  min-height: 300px;
}

.spinner {
  animation: rotate 2s linear infinite;
  z-index: 2;
  width: 50px;
  height: 50px;
}
.inverted .spinner {
  animation: rotate 2s linear infinite reverse;
}

.spinner_path {
  stroke-linecap: round;
  animation: dash 1.5s ease-in-out infinite;
}
.inverted .spinner_path {
  animation: dash 1.5s ease-in-out infinite reverse;
}

.spinner_container.primary .spinner_path {
  stroke: var(--ft-primary-color);
}
.spinner_container.secondary .spinner_path {
  stroke: var(--ft-secondary-color);
}
.spinner_container.tertiary .spinner_path {
  stroke: var(--ft-tertiary-color);
}
.spinner_container.green .spinner_path {
  stroke: var(--ft-success-color);
}
.spinner_container.red .spinner_path {
  stroke: var(--ft-error-color);
}
.spinner_container.yellow .spinner_path {
  stroke: var(--ft-warning-color);
}
.spinner_container.blue .spinner_path {
  stroke: var(--ft-info-color);
}

.spinner_container.primary {
  color: var(--ft-primary-color);
}
.spinner_container.secondary {
  color: var(--ft-secondary-color);
}
.spinner_container.tertiary {
  color: var(--ft-tertiary-color);
}
.spinner_container.green {
  color: var(--ft-success-color);
}
.spinner_container.red {
  color: var(--ft-error-color);
}
.spinner_container.yellow {
  color: var(--ft-warning-color);
}
.spinner_container.blue {
  color: var(--ft-info-color);
}

@keyframes rotate {
  100% {
    transform: rotate(360deg);
  }
}

@keyframes dash {
  0% {
    stroke-dasharray: 1, 150;
    stroke-dashoffset: 0;
  }
  50% {
    stroke-dasharray: 90, 150;
    stroke-dashoffset: -35;
  }
  100% {
    stroke-dasharray: 90, 150;
    stroke-dashoffset: -124;
  }
}
