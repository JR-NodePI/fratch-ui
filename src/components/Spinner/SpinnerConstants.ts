export const SpinnerType = {
  PRIMARY: 'primary',
  SECONDARY: 'secondary',
  TERTIARY: 'tertiary',
  RED: 'red',
  BLUE: 'blue',
  GREEN: 'green',
  YELLOW: 'yellow',
} as const;
