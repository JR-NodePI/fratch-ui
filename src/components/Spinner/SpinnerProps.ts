export const SpinnerType = Object.freeze({
  PRIMARY: 'primary',
  SECONDARY: 'secondary',
  TERTIARY: 'tertiary',
  RED: 'red',
  BLUE: 'blue',
  GREEN: 'green',
  YELLOW: 'yellow',
});

export type SpinnerProps = {
  cover?: boolean;
  inverted?: boolean;
  label?: string;
  type?: (typeof SpinnerType)[keyof typeof SpinnerType];
};
