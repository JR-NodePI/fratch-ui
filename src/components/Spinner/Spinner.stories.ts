import Spinner from './Spinner';
import type { Meta, StoryObj } from '@storybook/react';

const meta = {
  title: 'Example/Spinner',
  component: Spinner,
  tags: ['autodocs'],
  argTypes: {},
} satisfies Meta<typeof Spinner>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Docs: Story = {
  args: {},
};
