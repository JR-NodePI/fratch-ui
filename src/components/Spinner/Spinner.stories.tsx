import type { Meta, StoryObj } from '@storybook/react';

import Spinner from './Spinner';
import { SpinnerProps, SpinnerType } from './SpinnerProps';

const StoryRenderer = (args: SpinnerProps) => (
  <div>
    <Spinner {...args} />
    <h1>Fake page</h1>
    <p>
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam eros
      augue, fringilla quis nunc sit amet, cursus varius purus. Pellentesque non
      risus magna. Integer pulvinar odio augue, non gravida tortor ultricies eu.
      Nunc eget tincidunt nulla, eget consectetur enim. Donec pellentesque odio
      non massa consectetur, sodales finibus elit congue. Vivamus tincidunt,
      neque nec consequat sagittis, nibh lacus luctus metus, non lobortis est
      massa ac nulla. Donec placerat, risus ut convallis cursus, quam metus
      auctor turpis, pretium porttitor leo diam eget tellus. Nulla ultrices, leo
      sit amet facilisis auctor, arcu dui dignissim orci, eu tempor nulla odio
      sed orci. Pellentesque tempor posuere vehicula. Quisque vestibulum
      venenatis velit, eget facilisis dui sodales sed. Suspendisse quis est dui.
      Donec hendrerit, dolor eu vestibulum eleifend, lorem felis dictum sem,
      quis fringilla nisi est a lectus. Cras interdum, lacus sed vehicula
      feugiat, eros risus eleifend ipsum, non vestibulum quam urna quis enim.
      Vivamus ut est nulla.
    </p>
    <p>
      Curabitur ut tortor sit amet erat faucibus auctor at sed nunc. Donec
      posuere enim et nisi faucibus, tempor rutrum magna gravida. Duis maximus
      condimentum dui, eu malesuada tellus imperdiet vel. Curabitur viverra orci
      vel dui malesuada maximus. Orci varius natoque penatibus et magnis dis
      parturient montes, nascetur ridiculus mus. Vivamus ut ultricies eros, at
      accumsan sapien. Donec ornare sagittis ipsum, eget molestie nulla volutpat
      a. Morbi at mauris elementum, dapibus erat id, accumsan sapien. Praesent
      condimentum consequat quam, a ultrices nunc maximus in. Etiam tempus eros
      dui, ac egestas velit semper sollicitudin. Proin eget mattis diam.
    </p>
  </div>
);

const meta = {
  title: 'Example/Spinner',
  component: StoryRenderer,
  tags: ['autodocs'],
  argTypes: {
    type: {
      options: Object.values(SpinnerType),
      control: { type: 'select' },
    },
  },
} satisfies Meta<typeof Spinner>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Docs: Story = {
  args: {
    cover: false,
    type: 'primary',
    inverted: false,
    label: 'Loading...',
  },
};
