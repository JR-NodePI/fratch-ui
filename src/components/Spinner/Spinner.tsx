import { c } from '../../helpers/classNameHelpers';
import { SpinnerType } from './SpinnerConstants';
import { type SpinnerProps } from './SpinnerProps';

import styles from './Spinner.module.css';

const Spinner = ({
  cover,
  inverted,
  label,
  type = SpinnerType.PRIMARY,
}: SpinnerProps): JSX.Element => {
  return (
    <div
      className={c(
        styles.spinner_container,
        cover ? styles.cover : '',
        inverted ? styles.inverted : '',
        styles[type]
      )}
    >
      <svg className={c(styles.spinner)} viewBox="0 0 50 50">
        <circle
          className={c(styles.spinner_path)}
          cx="25"
          cy="25"
          r="20"
          fill="none"
          strokeWidth="5"
        ></circle>
      </svg>
      {label && <h5>{label}</h5>}
    </div>
  );
};

export default Spinner;
