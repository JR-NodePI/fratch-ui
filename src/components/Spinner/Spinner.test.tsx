import { render } from '@testing-library/react';
import { describe, expect, it } from 'vitest';

import Spinner from './Spinner';
import { type SpinnerProps } from './SpinnerProps';

describe('Spinner', () => {
  const setup = (props: SpinnerProps = {}): ReturnType<typeof render> =>
    render(<Spinner {...props} />);

  it('should render properly', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it('should render with cover', () => {
    const { container } = setup({
      cover: true,
    });
    expect(container).toMatchSnapshot();
  });

  it('should render with inverted', () => {
    const { container } = setup({
      inverted: true,
    });
    expect(container).toMatchSnapshot();
  });

  it('should render with label', () => {
    const { container } = setup({
      label: 'mock label',
    });
    expect(container).toMatchSnapshot();
  });

  it('should render with type', () => {
    const { container } = setup({
      type: 'secondary',
    });
    expect(container).toMatchSnapshot();
  });
});
