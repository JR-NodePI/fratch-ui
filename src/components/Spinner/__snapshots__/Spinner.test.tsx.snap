// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Spinner > should render properly 1`] = `
<div>
  <div
    class="_spinner_container_83c47e _primary_83c47e"
  >
    <svg
      class="_spinner_83c47e"
      viewBox="0 0 50 50"
    >
      <circle
        class="_spinner_path_83c47e"
        cx="25"
        cy="25"
        fill="none"
        r="20"
        stroke="currentColor"
        stroke-width="5"
      />
    </svg>
  </div>
</div>
`;

exports[`Spinner > should render with className 1`] = `
<div>
  <div
    class="_spinner_container_83c47e _primary_83c47e"
  >
    <svg
      class="_spinner_83c47e"
      viewBox="0 0 50 50"
    >
      <circle
        class="_spinner_path_83c47e"
        cx="25"
        cy="25"
        fill="none"
        r="20"
        stroke="currentColor"
        stroke-width="5"
      />
    </svg>
  </div>
</div>
`;

exports[`Spinner > should render with cover 1`] = `
<div>
  <div
    class="_spinner_container_83c47e _cover_83c47e _primary_83c47e"
  >
    <svg
      class="_spinner_83c47e"
      viewBox="0 0 50 50"
    >
      <circle
        class="_spinner_path_83c47e"
        cx="25"
        cy="25"
        fill="none"
        r="20"
        stroke="currentColor"
        stroke-width="5"
      />
    </svg>
  </div>
</div>
`;

exports[`Spinner > should render with inverted 1`] = `
<div>
  <div
    class="_spinner_container_83c47e _inverted_83c47e _primary_83c47e"
  >
    <svg
      class="_spinner_83c47e"
      viewBox="0 0 50 50"
    >
      <circle
        class="_spinner_path_83c47e"
        cx="25"
        cy="25"
        fill="none"
        r="20"
        stroke="currentColor"
        stroke-width="5"
      />
    </svg>
  </div>
</div>
`;

exports[`Spinner > should render with label 1`] = `
<div>
  <div
    class="_spinner_container_83c47e _primary_83c47e"
  >
    <svg
      class="_spinner_83c47e"
      viewBox="0 0 50 50"
    >
      <circle
        class="_spinner_path_83c47e"
        cx="25"
        cy="25"
        fill="none"
        r="20"
        stroke="currentColor"
        stroke-width="5"
      />
    </svg>
    <h5>
      mock label
    </h5>
  </div>
</div>
`;

exports[`Spinner > should render with type 1`] = `
<div>
  <div
    class="_spinner_container_83c47e _secondary_83c47e"
  >
    <svg
      class="_spinner_83c47e"
      viewBox="0 0 50 50"
    >
      <circle
        class="_spinner_path_83c47e"
        cx="25"
        cy="25"
        fill="none"
        r="20"
        stroke="currentColor"
        stroke-width="5"
      />
    </svg>
  </div>
</div>
`;
